### PR TITLE
 Fix issue that EntityUpdateAction increments version despite veto on update

### LINKED ISSUE #3198: EntityUpdateAction increments version despite veto on update
The _EntityUpdateAction_ requests whether the update is vetoed by calling _preUpdate()._ Even if the returned value is true (the update shall NOT be executed), the _nextVersion_ is set to the actual entity, only executing the update itself is skipped. This may affect Optimistic Locking.

The corresponding code is
```c#
public override void Execute(){
    ...
    bool veto = PreUpdate();
    ...
    if (!veto){
	persister.Update(id, state, dirtyFields, hasDirtyCollection, previousState, previousVersion, instance, null, session);
    }
    ...
    if (entry.Status == Status.Loaded || persister.IsVersionPropertyGenerated) {
	...
	// have the entity entry perform post-update processing, passing it the
	// update state and the new version (if one).
	entry.PostUpdate(instance, state, nextVersion);
    }
    ...
}
```

where _nextVersion_ is the value of the _Version_ property previously had plus one. Either the _nextVersion_ should be the same as the previous value or the _entry.postUpdate(...)_ must not be executed if veto is true.

I am working for PR for it.

### PR DESCRIPTION
**Pull request by chenhh021**
Fix #3198. The fix and test are straightforward.

### CODE REVIEW

File: src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+namespace NHibernate.Test.NHSpecificTest.GH3198
+{
+	/// <summary>
+	/// Fixture using 'by code' mappings
+	/// </summary>
+	/// <remarks>
+	/// This fixture is identical to <see cref="Fixture" /> except the <see cref="Entity" /> mapping is performed 
+	/// by code in the GetMappings method, and does not require the <c>Mappings.hbm.xml</c> file. Use this approach
+	/// if you prefer.
+	/// </remarks>
```
Comment (hazzik) [inline src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs:30]:
```suggestion
```

File: src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+				await (session.FlushAsync());
+
+				var versionAfterflush = entity.Version;
+				
+				Assert.That(versionAfterflush, Is.EqualTo(versionBeforeFlush), "The entity version must not change when update is vetoed");
+			}
+		}
+		
+		// A listener always returning true
+		public partial class testEventListener : IPreUpdateEventListener
```
Comment (hazzik) [inline src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs:103]:
```suggestion
		public partial class TestEventListener : IPreUpdateEventListener
```

File: src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	/// <remarks>
+	/// This fixture is identical to <see cref="Fixture" /> except the <see cref="Entity" /> mapping is performed 
+	/// by code in the GetMappings method, and does not require the <c>Mappings.hbm.xml</c> file. Use this approach
+	/// if you prefer.
+	/// </remarks>
+	[TestFixture]
+	public partial class ByCodeFixtureAsync : TestCaseMappingByCode
+	{
+		private static readonly int EXAMPLE_ID_VALUE = 1;
+		private readonly testEventListener listener = new testEventListener();
```
Comment (hazzik) [inline src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs:35]:
```suggestion
		private readonly TestEventListener listener = new TestEventListener();
```

File: src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+			base.Configure(configuration);
+		}
+		protected override HbmMapping GetMappings()
+		{
+			var mapper = new ModelMapper();
+			mapper.Class<Entity>(rc =>
+			{
+				rc.Table("Entity");
+				rc.Id(x => x.id, m => m.Generator(Generators.Assigned));
+				rc.Property(x => x.name, x=>x.Type<StringType>());
```
Comment (hazzik) [inline src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs:53]:
```suggestion
				rc.Property(x => x.name, x => x.Type<StringType>());
```

File: src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+			using (var transaction = session.BeginTransaction())
+			{
+				session.CreateQuery("delete from System.Object").ExecuteUpdate();
+
+				transaction.Commit();
+			}
+		}
+
+		[Test]
+		public async Task testVersionNotChangedWhenPreUpdateEventVetodAsync()
```
Comment (hazzik) [inline src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs:83]:
```suggestion
		public async Task TestVersionNotChangedWhenPreUpdateEventVetodAsync()
```

File: src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+				var versionBeforeFlush = entity.Version;
+				
+				await (session.FlushAsync());
+
+				var versionAfterflush = entity.Version;
+				
+				Assert.That(versionAfterflush, Is.EqualTo(versionBeforeFlush), "The entity version must not change when update is vetoed");
+			}
+		}
+
```
Comment (hazzik) [inline src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs:101]:
```suggestion
```

File: src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+			public Task<bool> OnPreUpdateAsync(PreUpdateEvent @event, CancellationToken cancellationToken)
+			{
+				return Task.FromResult<bool>(true);
+			}
+			public bool OnPreUpdate(PreUpdateEvent @event)
+			{
+				return true;
+			}
+			
+		}
```
Comment (hazzik) [inline src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs:119]:
```suggestion
		}

```

Comment (hazzik) [inline src/NHibernate.Test/Async/NHSpecificTest/GH3198/FixtureByCode.cs:119]:
```suggestion
		}

```

**Comment by hazzik** [comment]
Looks alright.

**Review by hazzik — APPROVED** [review]

**Comment by bahusoid** [comment]
Port of https://github.com/hibernate/hibernate-orm/pull/3699

**Pull request merged**